File: 0285-inorder-successor-in-bst/0285-inorder-successor-in-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    TreeNode ans = null;
    TreeNode prev = null;
    void inOrder(TreeNode node,TreeNode p) {
        if(node == null) return;
        
        inOrder(node.left,p);
        
        if(prev != null && prev == p) ans = node;
        
        
        prev = node;
        inOrder(node.right,p);
    }
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        inOrder(root,p);
        return ans;
    }
}

File: 0285-inorder-successor-in-bst/0285-inorder-successor-in-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* ans = NULL;
    TreeNode* prev = NULL;
    void inOrder(TreeNode* node, TreeNode* p) {
        if(!node) return;
        
        inOrder(node->left,p);
        
        if(prev && prev == p) ans = node;
        
        prev = node;
        inOrder(node->right,p);
    }
    TreeNode* inorderSuccessor(TreeNode* root, TreeNode* p) {
        inOrder(root,p);
        return ans;
    }
};